### PR TITLE
Universal thresholding

### PR DESCRIPTION
**Pull request by abruefach**
Disk Detection -
find_bragg_disks_serial- added universal thresholding and pointlistarray histogram capabilities

Universal_Thresholding - added ability to apply universal thresholding to any pointlistarray

get_pointlistarray_hist - added ability to retrieve histogram counts and bins from pointlistarray

Visualize-
show_hist- added function to return histogram of any ndarray input

### CODE REVIEW

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:615]: suggest adding (optional)

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:601]: remove blank line

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:620]: remove blank line

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:624]: remove blank line

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:631]: remove blank line

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:628]: remove blank line

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:639]: remove blank line

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:641]: remove blank line

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:644]: rm

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:646]: rm

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:660]: make double lines single lines, remove blank lines

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:691]: rm

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:707]: rm

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:715]: rm

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:718]: rm

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:720]: rm

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:725]: rm

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:739]: rm blank lines

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:690]: I'd suggest we add a `plot_hist` flag as well, which defaults to `True`, so that users have the option of (a) JUST getting an array of all the peak intensities, (b) getting that array and plotting it's histogram (default), or (c) getting the array, and getting the histogram values

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:705]: Suggest
```
counts        (ndarray) returned iff return_hist==True. The histogram bin counts
bin_edges  (ndarray) returned iff return_hist==True. The histogram bin edges
```

I would also strongly suggest we rename `bin_values` to either `bin_edges` or `bin_centers`, depending on which we're returning.  Right now it seems we're returning the edges (which is good, this is what numpy users will be used to / expect, even if this convention is sometimes frustrating / dumb.  Please see my comment in the visualize function!)

File: py4DSTEM/visualize/visualize.py
Comment (bsavitzky) [inline py4DSTEM/visualize/visualize.py:418]: rm

File: py4DSTEM/visualize/visualize.py
Comment (bsavitzky) [inline py4DSTEM/visualize/visualize.py:417]: suggest we change the last kwarg to `returnfig` for consistency with the other vis functions, where if this flag is True the figure is returned, and if False it is plotted and not returned

File: py4DSTEM/visualize/visualize.py
Comment (bsavitzky) [inline py4DSTEM/visualize/visualize.py:436]: This is actually not quite right!  np.histogram returns the bin *edges*.  So your scatterplot is showing histogram values that are all offset to the left by 1/2 the bin width.  Actually, I'm surprised this isn't breaking, as bin_values and counts should have lengths that differ by one...so perhaps I'm missing something here?  Normally to plot this histogram with the values, I'd suggest doing something like
```
counts,bin_edges = np.histogram(...)
bin_width = bin_edges[1]-bin_edges[0]
bin_centers = bin_edges[:-1]+bin_width/2.

fig,ax = plt.subplots()
ax.bar(bin_centers,counts,width=bin_width,align='center')
...
```
Strictly you don't need to specify the align param, as center is default, but explicitness is nice :3

**Review by bsavitzky — CHANGES_REQUESTED** [review]
Great work Alex!  I made a lot of comments, hopefully this is not too much (?!).  This is my first PR code review, so, we're all learning, I hope.  (!!)

The biggest two things are:
(1) I believe the histogram function in the visualize module is not quite right, and
(2) I'd suggest removing plotting the histogram from the main find_bragg_disks function.  My thought here is that this is a slow/bottleneck function, so I would encourage users to separate out running the compute intensive analysis code from examining / deciding about the threshold.  It means users will need two lines of code now - they'll need to subsequently run your `get_pointlistarray_hist` function - and I think that's probably ok here.  That said, I'm open to pushback - please do lmk what you think!

Overall looks great.  I'm submitting this as a 'Request changes' - my understanding is that at this point you should be able to make any edits you like to this same branch, and push them normally to your fork, and I should then see those updates in the PR.  Hopefully that will work smoothly?!

Ok!  Thanks for these new features!  These are going to be very helpful!!

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (abruefach) [reply py4DSTEM/process/diskdetection/diskdetection.py:371]: Did you want me to remove the universal threshold as well, or just  the histogram functionality? I'm fine with either. My workflow looked like the following:

find_bragg_disks_serial()
get_pointlistarray_hist()
universal_threshold()

However if people are working on several datasets from the same dat it might be helpful to just apply the same universal threshold during disk detection. I'm fine with whether you want the UT incorporated into the serial detection or not. I don't know if it will make a huge difference to users.

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (abruefach) [reply py4DSTEM/process/diskdetection/diskdetection.py:285]: Agreed, I'm not sure exactly how to consolidate right now. The parameters for global and relative will probably be different for users. I will think about it a bit more and see if there is a better solution.

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:703]: I think this line is redundant - if we already passed the previous `if not return_hist` line, we shouldn't need to check this again

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:706]: I'm not sure why we're returning the figure here.  Don't you just want to plot it?

File: py4DSTEM/visualize/visualize.py
Comment (bsavitzky) [inline py4DSTEM/visualize/visualize.py:423]: I think we're missing `returnfig` in the input descriptions

**Review by bsavitzky — CHANGES_REQUESTED** [review]
Looks excellent.  I added just a few more minor comments.

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [reply py4DSTEM/process/diskdetection/diskdetection.py]: Hmm.  I'm not certain...but I was thinking remove it here, so that users are required to use the workflow you describe above.  It might actually be quite helpful to break it down even a little more, such that the user can look at the results of setting some threshold before actually applying it.  I would imagine something like tagging which peaks will be removed without actually removing them, combined with a vis function that displays only the untagged peaks, or something like this.  But this is really a suggestion for a later feature, I don't want to hold this up now.

File: py4DSTEM/process/diskdetection/diskdetection.py
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:703]: is there an indentation error here?

**Comment by bsavitzky** [comment]
Thanks Alex!  We can try to loop back to consolidating the input arguments at some point in the future.

**Comment by bsavitzky** [comment]
Ok!  More news.

After merging this into the main dev and pulling into my current working branch to make use of some of the new functions (ty!), I found there were some import errors.  Not sure why they didn't come up on your machine!  No worries though, they were quick fixes, but, in order to avoid circular imports without causing too much headache, I removed the plot_hist option entirely from the diskdection functions.  I then figured that the real utility of the get_pointlistarray_hist function was getting all the intensities in a single array - the last part of making a histogram is easy, and can be done elsewhere.  So....I changed this function's name to get_pointlistarray_intensities, and simplified it to only accept a single pointlistarray.  If you want the histogram from here, you can pass this function's output to your visualize.show_hist function.  I then added an option to that function to return the histogram counts and bin_edges, so all the functionality you built is preserved, just shuffled around a little.  I think this is for the best - the code stays better organized this way, at the expense of the user needing to call the show_hist function in a new line of code instead of just passing a flag to get_pointlistarray_hist.  Last, I modified visualize.show_hist so it can also now add vertical lines to the plot, as this is a frequent need.

All these edits are presently living in the coords branch, as I am basically living there right now - if you need any of this, just grab the coords branch from the main repo.  I'll get it merged into dev as soon as possible.  Please lmk if you have any thoughts or concerns, we can of course still edit all this before it gets pushed into dev!

**Pull request merged**
